### ISSUE TITLE
Some labels are missing ?

### ISSUE THREAD

**Issue by bonswouar** [marcelklehr/recognize#37]
I was playing with the code when I noticed some labels are missing in the yml, probably on purpose ?
For example :
```yml
pizza:
  threshold: 0.2
  categories:
    - food
 ```
With this example the result is that pizza images are tagged as "food" only.
Although I feel like most of these (pizza specially) should be tagged too, people tend to search for those, don't you think ?

**Comment by marcelklehr** [comment]
With the example you give, it should actually add the label `pizza`, too.

**Comment by bonswouar** [comment]
> With the example you give, it should actually add the label `pizza`, too.

My bad, I c/p my own corrected "pizza" example instead of the one from the project, this should make more sense now ;)

**Comment by marcelklehr** [comment]
Ah, yes.

I think I removed the label because some pasta dishes were classified as pizza. Not sure if that was the best decision. I think this way we make the most out of the model's predictions. Ideally we'd have both labels with different thresholds, but I'm not sure that's possible with this structure.

**Comment by bonswouar** [comment]
I wouldn't mind tweaking it a bit to see if I can improve the thresholds (for ̀pizza` and others).

Do you have more tests that maybe you can share to be sure I won't break any other keyword ?
If you can join the expected results it would be perfect (It seems to be missing with the current images in `test/res`)

**Comment by marcelklehr** [comment]
There is no real test suite atm, sadly. If you're interested, I'd be happy to accept your contributions :)

**Comment by marcelklehr** [comment]
I'm also wondering if there are any better models out there that we could use.

**Comment by bonswouar** [comment]
For a real test suite we'd probably need to host the images somewhere else and retrieve them with the test script, as the more images we add the safer the tests will be.

About the models, the choices seem pretty limited with tfjs-models.. But you probably know more about this than me, do you have any other alternative in mind ?

**Comment by marcelklehr** [comment]
> For a real test suite we'd probably need to host the images somewhere else and retrieve them with the test script, as the more images we add the safer the tests will be.

Good idea! Maybe we can even use publicly hosted images from flickr/unsplash/etc

> About the models, the choices seem pretty limited with tfjs-models.. But you probably know more about this than me, do you have any other alternative in mind ?

Well, there is a way to convert normal tensorflow models to tf.js. But I haven't done enough research at this point to know which model is the most accurate.

**Comment by bonswouar** [comment]
> Good idea! Maybe we can even use publicly hosted images from flickr/unsplash/etc

Yeah totally, although I guess it would be better to use only Public Domain licensed images just to be safe _([Public Domain Dedication (CC0)](https://www.flickr.com/creativecommons/pdm-1.0/) and [Public Domain Work](https://www.flickr.com/creativecommons/cc0-1.0/) on Flickr)_.

> Well, there is a way to convert normal tensorflow models to tf.js. But I haven't done enough research at this point to know which model is the most accurate.

I had a quick look and found that EfficientNet (specially EfficientNet-Lite4) is the new go-to https://blog.tensorflow.org/2020/03/higher-accuracy-on-vision-models-with-efficientnet-lite.html

Although when you [compare them with latest updates](https://sotabench.com/benchmarks/image-classification-on-imagenet) it's not that clear which one would be better
 :
- EfficientNet-B3 (trained with [NoisyStudent](https://github.com/google-research/noisystudent)) seems to be more accurate than EfficientNet-Lite4  while taking 30% less time per image (about the same than MobileNetV2)
-  EfficientNet-B2 (with NoisyStudent) is between both but is twice as fast as EfficientNet-Lite4 .

I guess EfficientNet-Lite4 is still much more lightweight, but do we need that much efficiency? We're not supposed to run it on a mobile CPU anyway

**Comment by marcelklehr** [comment]
> I guess it would be better to use only Public Domain licensed images just to be safe

Good point! I agree.

> I've just noticed that the PR for a port of EfficientNet to tjjs-models has been abandoned... 

Ah, that's shame.

It seems someone else has already gone through the trouble of converting some python model to tf.js: https://github.com/ntedgi/node-efficientnet

**Comment by marcelklehr** [comment]
Testing should be quite simple. We could write a github flow downloading a list of image URLs into a temporary folder, then pass the list of files to the classifier and diff the output with an expected output (outputs are usually ordered).

**Comment by bonswouar** [comment]
> It seems someone else has already gone through the trouble of converting some python model to tf.js: https://github.com/ntedgi/node-efficientnet

Good find!! Although I see he didn't include EfficientNet-Lite, but that's worth a try : if it works out of the box with EfficientNet-B3 and we see it's not too much of a heavy load we could probably directly use it? Otherwise we could adapt some parts of this code yeah!

I'm wondering though, I don't know much about pre-trained model, it seems that checkpoints are basically backups of a state of a pre-trained model right? Then how would you add a checkpoint? I see this project has B0 to B7 but no NoisyStudent (which seems to get much better results)

> Testing should be quite simple. We could write a github flow downloading a list of image URLs into a temporary folder, then pass the list of files to the classifier

Yeah totally!
> and diff the output with an expected output (outputs are usually ordered).

If they're not always ordered (I guess it depends of the process time of each image?) then it wouldn't hurt to add a bit of logic in this validation though :)

**Comment by marcelklehr** [comment]
> I see this project has B0 to B7 but no NoisyStudent

Yep, good point. We'd need a noisystudent checkpoint if we want to use that: github.com/ntedgi/node-efficientnet/issues/59

I've started with a pretty naive testing scheme for now: https://github.com/marcelklehr/recognize/runs/3182099396
